jardin/instrumentation/notifier.py

```diff
@@ -0,0 +1,45 @@
+import uuid
+from threading import Lock
+from jardin import config as config
+from jardin.instrumentation.base_subscriber import BaseSubscriber
+from jardin.instrumentation.event import Event
+
+class NullNotifier:
+    def subscribe(self, _handler: BaseSubscriber) -> uuid.UUID:
+        return uuid.uuid4()
+
+    def unsubscribe(self, _subscriber_id: uuid.UUID) -> None:
+        pass
+
+    def report_event(self, ev: Event) -> None:
+        pass
+
+
+class SimpleNotifier:
+    def __init__(self):
+        self._subscribers = {}
+        self.lock = Lock()
+
+    def subscribe(self, handler: BaseSubscriber) -> uuid.UUID:
+        with self.lock:
+            subscriber_id = uuid.uuid4()
+            if not isinstance(handler, BaseSubscriber):
+                raise TypeError("Handler must be subclass of BaseSubscriber")
+            self._subscribers[subscriber_id] = handler
+            return subscriber_id
+
+    def unsubscribe(self, subscriber_id: uuid.UUID) -> None:
+        with self.lock:
+            del self._subscribers[subscriber_id]
+
+    def report_event(self, ev: Event) -> None:
+        with self.lock:
+            for subscriber_id in self._subscribers:
+                subscriber = self._subscribers[subscriber_id]
+                try:
+                    subscriber.report_event(ev)
```

Comment:
Is it worth holding the lock while we make this callback? I don't expect the callers would need it to be serialized, and if they do, they can do the locking themselves.

Reply:
What's a clean way to do that? 
The one way I can think of is to hold the lock, make a copy of subscribers, release the lock and iterate on that list. 
Releasing the lock before reporting the event and reacquiring it after wards feels a bit awkward + it will allow modification of the subscribers list during iteration.

Reply:
Your way sounds fine (copy the list of subscribers). And, yes, I do realize that we're basically back to what you were doing at the beginning of the PR when I started nit-picking 😄